pallets/portfolio/src/lib.rs

```diff
@@ -352,6 +353,14 @@ decl_module! {
                     .filter(|(identity, _, number)| !Portfolios::contains_key(identity, number))
                     .for_each(|(identity, name, _)| NameToNumber::remove(identity, name));
             });
+            storage_migrate_on!(StorageVersion::get(), 2, {
+                Portfolios::iter()
+                    .filter(|(identity, _, number)| !NameToNumber::contains_key(identity, number))
```

Comment:
The filter should check that the storaged number in `NameToNumber` matches the number in `Portfolios`.  Also the `number` parameter here is the `name`.